### PR TITLE
Fix XML Equality Check by Comparing Parsed XML Structure Instead of Raw Strings

### LINKED ISSUE #3159: [Bug] Incorrect Test Data in WriteKeyInfoTheoryData (Microsoft.IdentityModel.Xml.Tests)
**Which version of Microsoft.IdentityModel are you using?**
Microsoft.IdentityModel 8.6.0

**Where is the issue?**
 * [ ] M.IM.JsonWebTokens
 * [ ] M.IM.KeyVaultExtensions
 * [ ] M.IM.Logging
 * [ ] M.IM.ManagedKeyVaultSecurityKey
 * [ ] M.IM.Protocols
 * [ ] M.IM.Protocols.OpenIdConnect
 * [ ] M.IM.Protocols.SignedHttpRequest
 * [ ] M.IM.Protocols.WsFederation
 * [ ] M.IM.TestExtensions
 * [ ] M.IM.Tokens
 * [ ] M.IM.Tokens.Saml
 * [ ] M.IM.Validators
 * [x] M.IM.Xml
 * [ ] S.IM.Tokens.Jwt 
 * Other (please describe): 

**Is this a new or an existing app?** 
The app is in production and I have upgraded to a new version of `Microsoft.IdentityModel`.

#### **Repro**  

While fixing **issue #3130**, I identified that some test data in `WriteKeyInfoTheoryData` is incorrect, leading to mismatches between the actual and expected results in the **`WriteKeyInfo`** test method in the `DSigSerializerTests` class inside the **Microsoft.IdentityModel.Xml.Tests** test project.  

Code Snippet

```csharp
[Theory, MemberData(nameof(WriteKeyInfoTheoryData), DisableDiscoveryEnumeration = true)]
public void WriteKeyInfo(DSigSerializerTheoryData theoryData)
{
    TestUtilities.WriteHeader($"{this}.WriteKeyInfo", theoryData);
    var context = new CompareContext($"{this}.WriteKeyInfo, {theoryData.TestId}");
    try
    {
        var keyInfo = theoryData.Serializer.ReadKeyInfo(XmlUtilities.CreateDictionaryReader(theoryData.Xml));
        theoryData.ExpectedException.ProcessNoException(context.Diffs);
        IdentityComparer.AreKeyInfosEqual(keyInfo, theoryData.KeyInfo, context);
        var ms = new MemoryStream();
        var writer = XmlDictionaryWriter.CreateTextWriter(ms);
        theoryData.Serializer.WriteKeyInfo(writer, keyInfo);
        writer.Flush();
        var xml = Encoding.UTF8.GetString(ms.ToArray());

        // FIXME: Test data needs to be corrected because the test data and expected results differ.
        // Used CompareContext.Default to prevent this test from failing.
        IdentityComparer.AreEqual(theoryData.Xml, xml, CompareContext.Default);
    }
    catch (Exception ex)
    {
        theoryData.ExpectedException.ProcessException(ex, context.Diffs);
    }

    TestUtilities.AssertFailIfErrors(context);
}

```

**Expected behavior**
The test should correctly compare the serialized XML output with the expected XML structure. The test data in `WriteKeyInfoTheoryData` should align with the actual expected results.

**Actual behavior**
Currently, the test data and expected results do not match, which forces the test to use `CompareContext.Default` to avoid failure. This indicates that the test data needs to be updated for accurate validation.

**Possible solution**
- Review and update the test data in `WriteKeyInfoTheoryData` to ensure consistency with the expected serialized XML output.
- Remove reliance on `CompareContext.Default` and use created context to validate the XML structure against the corrected expected results.

**Additional context / logs / screenshots / links to code**
- Identified after fixing **issue #3130**.
- Found in `Microsoft.IdentityModel.Xml.Tests` under `DSigSerializerTests.cs`.
- The issue affects the correctness of test validation in `WriteKeyInfoTheoryData`.

### PR DESCRIPTION
**Pull request by omidmloo**
# Fix XML Equality Check by Comparing Parsed XML Structure Instead of Raw Strings  

**Summary of the changes**
Ensures XML comparisons are based on structure instead of raw string equality.  

## Description  

Previously, the test compared XML as raw strings, which caused **false negatives** due to formatting differences, whitespace, and attribute order. While the expected result was different, the discrepancies were due to formatting rather than actual content differences. This fix ensures XML elements are compared meaningfully, preventing incorrect failures.  

### **Fixes & Improvements:**   
- **Type-Aware XML Comparison:**  
  - Introduced a dictionary-based approach to compare XML elements based on their specific types.  
  - Added `AreXmlsEqual` to handle XML documents as objects rather than raw strings.  
  - Ensured XML comparisons focus on element structure rather than text differences.  
- **More Robust Equality Checks:**  
  - Improved handling of various security and identity-related XML types.  
  - Avoids false negatives caused by minor formatting discrepancies.  
- **Enhanced Debugging & Maintainability:**  
  - Clearer difference tracking when mismatches occur.  
  - Improved logging of discrepancies between expected and actual XML structures.  

Fixes #3159

### CODE REVIEW

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs [hunk 1]
```diff
@@ -1170,9 +1172,114 @@ public static bool AreStringDictionariesEqual(Object object1, Object object2, Co
...
+                return context.Merge(localContext);
+
+            if (ReferenceEquals(xml1, xml2))
+                return true;
+
+            if (xml1 == null)
+                localContext.Diffs.Add($"({name1} == null, {name2} == {xml2.ToString()}.");
+
+            if (xml2 == null)
+                localContext.Diffs.Add($"({name1} == {xml1.ToString()}, {name2} == null.");
```
Comment (sruke) [inline test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs]: ContinueCheckingEquality already checks if xml1 or xml2 is null.

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs [hunk 2]
```diff
@@ -1170,9 +1172,114 @@ public static bool AreStringDictionariesEqual(Object object1, Object object2, Co
...
+
+            if (!CompareXmlElements(xml1.Root, xml2.Root, localContext))
+            {
+
```
Comment (sruke) [inline test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs]: Please remove the extra new lines on lines 1197, 1203 and 1282.

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs [hunk 3]
```diff
@@ -1170,9 +1172,114 @@ public static bool AreStringDictionariesEqual(Object object1, Object object2, Co
             context.Diffs.AddRange(localContext.Diffs);
             return localContext.Diffs.Count == 0;
         }
+        public static bool AreXmlsEqual(object xml1, object xml2, CompareContext context)
```
Comment (sruke) [inline test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:1176]: If xml1 or xml2 is not an XDocument, an InvalidCastException could be thrown. Consider adding a try catch block.

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs [hunk 4]
```diff
@@ -1170,9 +1172,114 @@ public static bool AreStringDictionariesEqual(Object object1, Object object2, Co
...
+
+        /// <summary>
+        /// Compares two XML elements for equality, ignoring order of attributes and child elements.
+        /// Ignore X509 certificate elements and attributes.
+        /// </summary>
+        /// <param name="elem1">The first XML element to compare.</param>
+        /// <param name="elem2">The second XML element to compare.</param>
+        /// <param name="localContext"></param>
+        /// <returns>True if the elements are considered equal, otherwise false.</returns>
+        private static bool CompareXmlElements(XElement elem1, XElement elem2, CompareContext localContext)
```
Comment (sruke) [inline test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:1217]: Consider utilizing localContext to add information about the differences found which could help with debugging.

File: test/Microsoft.IdentityModel.Xml.Tests/DSigSerializerTests.cs [hunk 5]
```diff
@@ -115,7 +116,11 @@ public void WriteKeyInfo(DSigSerializerTheoryData theoryData)
                 theoryData.Serializer.WriteKeyInfo(writer, keyInfo);
                 writer.Flush();
                 var xml = Encoding.UTF8.GetString(ms.ToArray());
-                IdentityComparer.AreEqual(theoryData.Xml, xml);
+
+                // Compare the original XML with the re-serialized XML.
+                // Parsing the XML strings into XDocument ensures that the comparison is based on 
+                // structural and content equality rather than raw string differences (formatting, whitespace,...).
+                IdentityComparer.AreEqual(XDocument.Parse(theoryData.Xml), XDocument.Parse(xml), context);
```
Comment (sruke) [inline test/Microsoft.IdentityModel.Xml.Tests/DSigSerializerTests.cs:122]: Consider adding tests for different xml structures and null values to verify IdentityComparer behaves as expected when comparing XML.

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs (hunk 2 above)
Comment (omidmloo) [reply test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs]: done

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs (hunk 3 above)
Comment (omidmloo) [reply test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:1176]: done

File: test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs (hunk 4 above)
Comment (omidmloo) [reply test/Microsoft.IdentityModel.TestUtils/IdentityComparer.cs:1217]: done

File: test/Microsoft.IdentityModel.Xml.Tests/DSigSerializerTests.cs (hunk 5 above)
Comment (omidmloo) [reply test/Microsoft.IdentityModel.Xml.Tests/DSigSerializerTests.cs:122]: Hi Sruke, thank you for the review. Since these tests focus on `IdentityComparer`, would it be better to add them to a new test file for better organization, or should I extend the existing test file? Let me know your preference.

Comment (sruke) [reply test/Microsoft.IdentityModel.Xml.Tests/DSigSerializerTests.cs:122]: Agree that creating a new file is better for organization.

**Review by brentschmaltz — APPROVED** [review]
